providers/firestore.js

```diff
@@ -0,0 +1,183 @@
+13. Download the `firebase-admin` module.
+*/
+
+const { Provider } = require('klasa');
+const firebase = require('firebase-admin');
+const { FieldValue } = firebase;
+
+module.exports = class extends Provider {
+
+	constructor(...args) {
+		super(...args);
+		this.db = null;
+	}
+	/**
+	 * Initiaizes the firestore
+	 * @private
+	 */
+	async init() {
+		await firebase.initializeApp({
+			credential: firebase.credential.cert(this.client.options.providers.firestore.credentials),
+			databaseURL: this.client.options.providers.firestore.databaseURL
+		});
+
+		this.db = firebase.firestore();
+	}
+	/**
+	 *Checks if the table/collection exists
+	 * @param {string} table The name of the table/collection to be checked for existance
+	 * @returns {Promise<boolean>}
+	 */
+	hasTable(table) {
+		return this.db.collection(table).get().then(col => !!col.size);
+	}
+
+	/**
+	 * This method is a dummy method, as new tables/collections are automatically created by firestore when needed.
+	 * @param {string} table Name of the collection to create.
+	 * @returns {Promise<Collection>}
+	 */
+	createTable(table) {
+		return this.db.collection(table);
+	}
+
+	/**
+	 * Returns all the documents of a collection
+	 * @param {string} table Name of the collection, for which all values is to be returned
+	 * @returns {Promise<Array>}
+	 */
+	getAll(table) {
+		return this.db.collection(table).get().then(snaps => snaps.docs.map(snap => snap.data()));
+	}
+	/**
+	 * Returns all the Keys/Docs of a collection/tanle.
+	 * @param {string} table Name of the collection/table for which all Keys/Docs is to be returned.
+	 * @returns {Promise<Array>}
+	 */
+	getKeys(table) {
+		return this.db.collection(table).get().then(snaps => snaps.docs.map(snap => snap.id));
+	}
+	/**
+	 * Returns a particular Document from the table/Collection.
+	 * @param {string} table Name of the collection/Table to get the value from
+	 * @param {string} id Name of the Document/Key to be returned
+	 * @returns {Promise<Any>}
+	 */
+	get(table, id) {
+		return this.db.collection(table).doc(id).get();
+	}
+
+	/**
+	 * Checks if a collection has a particular Document/Key
+	 * @param {string} table Name of the collection/Table
+	 * @param {string} id Name of the Document/Key
+	 * @returns {Promise<boolean>}
+	 */
+	has(table, id) {
+		return this.db.collection(table).doc(id).get().then(data => data.exists);
+	}
+
+	/**
+	 * Creates a new Document/Field in a collection.
+	 * @param {string} table Name of the collection/table
+	 * @param {string} id Name of they Docment/Key
+	 * @param {(ConfigurationUpdateResultEntry[] | [string, any][] | Object<string, *>)} doc the object you want to insert in the table.
+	 * @returns {Promise<DocumentObject>}
+	 */
+	async create(table, id, doc = {}) {
+		return await this.db.collection(table).doc(id).set(this.parseUpdateInput(doc));
```

Comment:
Useless return-await, you can remove the async from the function aswell. Same for the methods below.

Reply:
Could you explain, why

Reply:
Because the promise from the method's content is awaited outside, you're awaiting the await when you await this method call.

It's exactly the same, just that you're doing `Promise<Promise<Object>>` (AsyncFunction + db call's promise) instead of `Promise<Object>`